packages/ember-glimmer/tests/integration/components/curly-components-test.js

```diff
@@ -917,4 +920,1044 @@ moduleFor('Components test: curly components', class extends RenderingTest {
+    assert.equal(this.$('#args-5').text(), 'Foo4Bar5Baz');
+    assert.equal(this.$('#helper').text(), 'Foo4Bar5Baz');
+
+    this.runTask(() => this.rerender());
+
+    assert.equal(this.$('#args-3').text(), 'Foo4Bar');
+    assert.equal(this.$('#args-5').text(), 'Foo4Bar5Baz');
+    assert.equal(this.$('#helper').text(), 'Foo4Bar5Baz');
+  }
+
+  ['@htmlbars arbitrary positional parameter conflict with hash parameter is reported']() {
+    this.registerComponent('sample-component', {
+      ComponentClass: Component.extend().reopenClass({
+        positionalParams: 'names'
+      }),
+      template: strip`
+        {{#each names as |name|}}
+          {{name}}
+        {{/each}}`
+    });
+
+    expectAssertion(() => {
+      this.render(`{{sample-component "Foo" 4 "Bar" names=numbers id="args-3"}}`, {
+        numbers: [1, 2, 3]
+      });
+    }, 'You cannot specify positional parameters and the hash argument `names`.');
+  }
+
+  ['@test can use hash parameter instead of arbitrary positional param [GH #12444]'](assert) {
+    this.registerComponent('sample-component', {
+      ComponentClass: Component.extend().reopenClass({
+        positionalParams: 'names'
+      }),
+      template: strip`
+        {{#each names as |name|}}
+          {{name}}
+        {{/each}}`
+    });
+
+    this.render('{{sample-component names=things}}', {
+      things: emberA(['Foo', 4, 'Bar'])
+    });
+
+    this.assertText('Foo4Bar');
+
+    this.runTask(() => this.rerender());
+
+    this.assertText('Foo4Bar');
+
+    this.runTask(() => this.context.get('things').pushObject(5));
+
+    this.assertText('Foo4Bar5');
+
+    this.runTask(() => this.context.get('things').shiftObject());
+
+    this.assertText('4Bar5');
+
+    this.runTask(() => this.context.get('things').clear());
+
+    this.assertText('');
+
+    this.runTask(() => this.context.set('things', emberA(['Foo', 4, 'Bar'])));
+
+    this.assertText('Foo4Bar');
+  }
+
+  ['@htmlbars can use hash parameter instead of positional param'](assert) {
+    this.registerComponent('sample-component', {
+      ComponentClass: Component.extend().reopenClass({
+        positionalParams: ['first', 'second']
+      }),
+      template: '{{first}} - {{second}}'
+    });
+
+    this.render(strip`
+      {{sample-component "one" "two" id="two-positional"}}
+      {{sample-component "one" second="two" id="one-positional"}}
+      {{sample-component first="one" second="two" id="no-positional"}}`);
+
+    assert.equal(this.$('#two-positional').text(), 'one - two');
+    assert.equal(this.$('#one-positional').text(), 'one - two');
+    assert.equal(this.$('#no-positional').text(), 'one - two');
+
+    this.runTask(() => this.rerender());
+
+    assert.equal(this.$('#two-positional').text(), 'one - two');
+    assert.equal(this.$('#one-positional').text(), 'one - two');
+    assert.equal(this.$('#no-positional').text(), 'one - two');
+  }
+
+  ['@skip dynamic arbitrary number of positional parameters'](assert) {
```

Comment:
I have added the `@htmlbars` to title and commented the lines for the U part of the helper. I can move the dynamic part to the proper test suite as well if you want (more on this on a following comment), though rest positional parameters is not working for the `component` helper at all.